Verify the fraud proof completely on the client side

liuchengxu opened the pull request:
This PR aims to replace the fraud proof externality with a pure client-side solution. The idea is to perform an additional pre-validation before calling the upstream `validate_transaction` once an extrinsic is submitted to the tx pool. In order to do this, a `BasicPool` wrapper is added. 

There are still a few TODOs, but most of the work has been done and can be reviewed, nice to have some early feedback.

- [x] Add a test.
    - Writing a test is not hard, but somehow I encountered a really weird issue while writing it: this function call [sp_state_machine::execution_proof_check](https://github.com/subspace/subspace/blob/2a56ab7654/crates/subspace-fraud-proof/src/lib.rs#L267) never returns and results in the test running forever(https://github.com/subspace/subspace/runs/6742525914?check_suite_focus=true). I need more time to figure it out. Update: still have no idea what happened exactly, but found a workaround.
- [x] Verify the fraud proof in consensus.
- [x] Remove the deprecated fraud proof externality.

liuchengxu committed:
Add runtime api: extract_fraud_proof()

liuchengxu committed:
Remove fraud proof ext

liuchengxu committed:
Initial pool module in subspace-service

No logic changes, this commit only makes a wrapper for the upstream `new_full` function.

liuchengxu committed:
Inline `sc_transaction_pool::BasicPool::new_full`

`BasicPool` and `FullChainApi` will be replaced with `BasicPoolWrapper`
and `FullChainApiWrapper` later.

liuchengxu committed:
Introduce `BasicPoolWrapper`

`BasicPoolWrapper` now is only a mere wrapper type of `BasicPool` without any
modifications.

liuchengxu committed:
Introduce `FullChainApiWrapper`

Similar to `BasicPoolWrapper`, `FullChainApiWrapper` now a wrapper type
of the upstream `FullChainApi` without any actual modifications.

liuchengxu committed:
Copy the original `submit_local` implementation

liuchengxu committed:
Implement `submit_local`

liuchengxu committed:
Rename C in `NewFull` to Client

liuchengxu committed:
Integrate FraudProofVerifier into the custom tx pool

Essentially the goal of our custom tx pool is to perform the extrinsic pre-validation before
calling the inner `validate_transaction` which calls into the runtime.

liuchengxu committed:
Add TODO for verifying the fraud proof in consensus and make clippy happy

We need to check the fraud proof in consensus otherwise the dishonest
farmer can alter the source code and include an invalid fraud proof
extrinsic in a block once the fraud proof verification in runtime is removed.

liuchengxu committed:
Add test for fraud proof verification in tx pool

liuchengxu committed:
Fetch execution wasm bundle earlier while verifying fraud proof

The previous version that fetch the runtime code inside `execution_proof_check`
causes a really weired issue that the underlying runtime api (https://github.com/subspace/subspace/blob/2072282b10/crates/subspace-fraud-proof/src/lib.rs#L193-L196)
seemingly is never able to return. The new version just works, but I
still have no clue what the root cause is.

liuchengxu committed:
Add new variant RuntimeApi in fraud proof verification error

nazar-pc commented on crates/subspace-service/src/pool.rs:
```diff
@@ -0,0 +1,36 @@
+
+pub(super) fn new_full<Block, Client>(
+    config: &Configuration,
+    spawner: impl SpawnEssentialNamed,
```
Try to not use `impl`, it causes inconveniences in some cases.

liuchengxu replied:
Updated, not sure how helpful it can be though as it's already used in the upstream `FullChainApi::new`.

nazar-pc commented on crates/subspace-service/src/pool.rs:
```diff
@@ -1,18 +1,199 @@
+        _at: &BlockId<Self::Block>,
+        _xt: sc_transaction_pool_api::LocalTransactionFor<Self>,
+    ) -> Result<Self::Hash, Self::Error> {
+        todo!("Impl submit_local")
```
Commit message says no modifications, but this `todo` indicates the opposite?

liuchengxu replied:
Okay, strictly speaking, that commit and then two more after it altogether makes it, this `todo!` is only used to make it compile. I meant no logical changes in the end.

nazar-pc commented on crates/subspace-service/src/pool.rs:
```diff
@@ -166,25 +193,22 @@ where
-            traits::SaturatedConversion, transaction_validity::TransactionValidityError,
-        };
+        use sp_runtime::traits::SaturatedConversion;
+        use sp_runtime::transaction_validity::TransactionValidityError;
```
Can these be moved to normal imports instead at the top of the file?

nazar-pc commented on crates/subspace-service/Cargo.toml:
```diff
@@ -21,6 +21,7 @@ derive_more = "0.99.17"
 frame-support = { version = "4.0.0-dev", git = "https://github.com/subspace/substrate", rev = "5f0aa1feb7250ac7b8c1b9928f87b2420b530e22" }
 futures = "0.3.21"
 jsonrpsee = { version = "0.13.1", features = ["server"] }
+pallet-executor = { version = "0.1.0", path = "../pallet-executor" }
```
We really shouldn't have `pallet-executor` anywhere in the service dependencies

liuchengxu replied:
I can understand service is generic about Runtime and ExecutorDispatch, why pallet can't be included for a specific project instead of a general framework? 

And if so, what's your suggestion on this https://github.com/subspace/subspace/blob/961104c4b5a846799adf17bd5d8af949d912d19c/crates/subspace-service/src/pool.rs#L131-L133

move `InvalidTransactionCode` to `sp-executor`?

nazar-pc replied:
Pallets are runtime logic. Even if there is an enum that looks similar to what you need in service, it is not a good reason to include pallet in the service directly.

liuchengxu replied:
Well, polkadot does use a few pallets in their service. I don't mind either way, if you prefer not to use pallet in the service, then we should move it into `sp-executor`, or you have any other suggestion?

nazar-pc replied:
Let's move it to `sp-executor` for now

nazar-pc commented on crates/subspace-fraud-proof/src/lib.rs:
```diff
@@ -294,6 +294,28 @@ where
 }
 
+/// Verify fraud proof.
+pub trait VerifyFraudProof {
```
Why do we need a trait for this?

liuchengxu replied:
With this trait, integrating the fraud proof verifier will be much easier, you don't have to bring the generics of the underlying `ProofVerifier` to the consumers in tx pool and consensus. Also useful for the tests where a real proper fraud proof verifier is not necessary.

nazar-pc commented on crates/subspace-service/src/pool.rs:
```diff
@@ -42,11 +45,14 @@ type ReadyIteratorFor<PoolApi> = BoxedReadyIterator<ExtrinsicHash<PoolApi>, Extr
 type PolledIterator<PoolApi> = Pin<Box<dyn Future<Output = ReadyIteratorFor<PoolApi>> + Send>>;
 
-pub struct FullChainApiWrapper<Block, Client> {
+pub struct FullChainApiWrapper<Block, Client, VerifierClient, Verifier> {
```
It looks like `Client` and `VerifierClient` are the same type. Do you expect them to ever be different from point of view of arguments supplied?

liuchengxu replied:
No, thanks for pointing out, it's a tech debt I forgot to revisit, updated.

nazar-pc commented on crates/subspace-service/src/pool.rs:
```diff
@@ -122,18 +122,24 @@ where
+        match self.client.runtime_api().extract_fraud_proof(at, &uxt) {
+            Ok(maybe_fraud_proof) => {
+                if let Some(fraud_proof) = maybe_fraud_proof {
+                    if let Err(err) = self.verifier.verify_fraud_proof(&fraud_proof) {
```
Shouldn't this be done in a spawned blocking task though? I'm worried it might block async executor for some time.

liuchengxu replied:
I'm unsure about it but sounds reasonable indeed, we don't yet know how heavy the fraud proof verification can be.

nazar-pc replied:
That is my worry: retrieving runtime code, preparing environment and running fraud proof can be potentially quite costly (especially when incorrect FP is generated in a specific way). So I'd better wrap it just to be safe.

liuchengxu replied:
Retrieving the runtime code and preparing the environment for verifying the fraud proof is no different from the regular operation, what I'm only uncertain about is the cost of running the fraud proof. Will take care of this after addressing the other comments.

liuchengxu replied:
Done in https://github.com/subspace/subspace/pull/551/commits/afb70084806bc83d968188e96ff6fd01c39acd48

nazar-pc commented on crates/subspace-fraud-proof/src/lib.rs:
```diff
@@ -252,11 +241,16 @@ where
             ..
         } = proof;
 
+        let at = PBlock::Hash::decode(&mut parent_hash.encode().as_slice())
```
If moving here resolves an issue, it seems to be like a race condition where transaction that generates runtime upgrade may not go through yet. I'd really like to figure out what is going on instead as we might hit the same issue under different circumstances and it feels important.

liuchengxu replied:
I don't think runtime upgrade is related because there is no runtime upgrade at all in the test. And we are using calling a runtime API which is called on top of the state that the certain block has been applied, the runtime upgrade must have been done at that point if there is one.

nazar-pc commented on crates/sp-executor/src/lib.rs:
```diff
@@ -296,6 +296,9 @@ pub enum VerificationError {
     /// Failed to decode the header produced by `finalize_block`.
     HeaderDecode(parity_scale_codec::Error),
+    /// Runtime api error.
+    #[cfg(feature = "std")]
```
Are you sure `VerificationError` enum should even be in this crate anymore? It doesn't seem to belong here if we're not doing verification in the runtime.

liuchengxu replied:
I thought that too initially, but still found one usage:
https://github.com/subspace/subspace/blob/961104c4b5a846799adf17bd5d8af949d912d19c/crates/sp-executor/src/lib.rs#L266-L269

`decode_execution_result` is now only used in the client, perhaps we can make it std only in sp-executor, then whole `VerificationError` can be std only. I don't have a strong preference as I don't think it's siginificant.

liuchengxu replied:
Okay, another reason to keep it in this crate https://github.com/subspace/subspace/pull/551/commits/961104c4b5a846799adf17bd5d8af949d912d19c#diff-c7fab4838df44656cad8198d607fd03699ad595255a9a5b767ed9487291f9322R253-R255 because I'm somewhat kind of with using sp-executor instead of subspace-fraud-proof in sc-consensus-subspace. Then the problem becomes if we want to make this error enum std only.

nazar-pc replied:
The question then is whether we should try to smash everything into this enum or we can have several separate enums and a way to convert one into another if necessary.

liuchengxu replied:
I don't quite get your last comment, what do you mean by several separate enums?

nazar-pc replied:
If there are different places that generate errors, then we can have multiple enums like this for errors. Even if they are similar, that is fine, we don't have to reuse enums if it is not a good fit.

liuchengxu committed:
Verify the fraud proof extrinsic while importing the block

liuchengxu committed:
Remove unnecessary generic VerifierClient

liuchengxu committed:
Not use `impl` and reorg the imports

nazar-pc commented on crates/sc-consensus-subspace/src/lib.rs:
```diff
@@ -1071,7 +1085,16 @@ where
         // and root blocks in the inherents are set correctly.
         if let Some(extrinsics) = extrinsics {
-            // TODO: Verify the fraud proof is correct if the extrinsic `submit_fraud_proof` is found.
+            // Ensure the fraud proof is valid if the `submit_fraud_proof` extrinsic is included.
```
I agree this'll work, but it seems to be completely orthogonal to consensus. What do you think about moving it to a different `BlockImport` in a separate crate, so we can keep these concerns separate?

liuchengxu replied:
Broadly speaking, I think it's related to the consensus as you're dependent on the global consensus between the nodes in the network, but I agree with moving it to `subspace-fraud-proof` which provides a `BlockImport` checking the fraud proof, which sounds clearer. Will update.

nazar-pc replied:
I was thinking of a separate `sc-` crate potentially

liuchengxu replied:
I don't have a strong opinion, then perhaps `sc-consensus-fraud-proof`?

nazar-pc replied:
Looks fine to me

liuchengxu committed:
Introduce sc-consensus-fraud-proof

This basically reverts commit 4b36c314abbc8281ec1469e52a22833a87fc7ee6
and replace it with a new crate `sc-consensus-fraud-proof`. Now
`sc-consensus-fraud-proof` provides a `BlockImport` which checks the
fraud proof in each block and is wrapped in `sc-consensus-subspace`.

liuchengxu committed:
Run the actual fraud proof verification in another blocking task

nazar-pc commented on crates/subspace-service/src/pool.rs:
```diff
@@ -121,16 +127,48 @@ where
+                    let fraud_proof_verifier = self.verifier.clone();
+                    let at = *at;
+
+                    return async move {
```
Just wondering, why not `return spawner.spawn_blocking` then, so we have less wrappers and don't need to clone spawner?

nazar-pc replied:
Return type doesn't work like that, sorry

liuchengxu committed:
Avoid using pallet-executor in service by moving the required code to sp-executor

liuchengxu committed:
Merge branch 'main' into client-side-fraud-proof-verification
